Add account linking to concepts (#240)
onflow/docs: docs/concepts/account-linking/index.mdx

sidebar_position: 15
---

# Account Linking

Account linking is a unique Flow concept that enables sharing ownership over accounts. In order to understand how we can achieve that we must first understand how accounts on Flow are accessed.

Each account on flow can be accessed in Cadence through two types, `PublicAccount` and `AuthAccount`. As the name implies the public account type gives access to all public account information such as address, balance, storage capacity, etc., but doesn’t allow changing the account, whereas the auth account type allows the same access the public account has but also allows changing the account, which includes changing the account keys, managing the deployed contracts, and managing account capabilities.

![Flow account structure](resources/account.png)

## Accessing AuthAccount

Accessing the auth account allows for account modification, so it's essential to safeguard this access by mandating that transactions are signed by the account being accessed. A transaction can list multiple auth accounts it wants to access as part of the `pre` section of the transaction. Read more about transaction signing in the transaction documentation.

> The idea of account ownership actually translates to the ability to access the auth account.

Reviewer: Maybe:
```suggestion
Since access to the `AuthAccount` object enables state change, the idea of account ownership actually translates to the ability to access the auth account. Traditionally, you might consider this the same as having key access on an account, but we'll see in just a minute how programmatic, ownership-level access is unlocked with Capabilities on Flow.
```

## AuthAccount Capabilities

Understanding [Cadence capabilities](../../cadence/language/capability-based-access-control) is required to follow this section.

Cadence allows the creation of capabilities to delegate access to account storage, meaning any account that has a valid capability to another account object in the storage can access it. This is a powerful feature on its own. The access to the object can be limited when creating a capability so only intended functions or fields can be accessed.

> What makes account linking possible is an extension of the capabilities. Similar to how storage capabilities allow gaining access to a value stored in an account's storage, AuthAccount capabilities allow gaining access to an AuthAccount.

Reviewer: ```suggestion
Account linking is made possible by the extension of Capabilities on the `AuthAccount` object itself. Similar to how storage capabilities allow access to a value stored in an account's storage, AuthAccount capabilities allow access to an AuthAccount.
```

### Creating Account Links

> Creating an AuthAccount capability and publishing that capability to another account is what we define as account linking. An account owning the `AuthAccount` capability of another account is called a parent account, and the account sharing the capability with the other account becomes its child account.

Reviewer: In this case, the child account would create the `AuthAccount` Capability and issue it to the parent. So maybe: 

```suggestion
When referring to 'account linking', we mean that an `AuthAccount` capability is created by one account and published to another account. Once claimed, the account claiming the published `AuthAccount` capability is the parent account while the publishing account is now the parent's child account.
```

These two steps are implemented in Cadence as two transactions:

************************************Create capability************************************

The account B creates and publishes the `AuthAccount` capability to the account A at the address `0x01`

```json
#allowAccountLinking

transaction {
    prepare(signer: AuthAccount) {
        let capability = signer.linkAccount(/private/accountCapA)!
        signer.inbox.publish(capability, name: "accountCapA", recipient: 0x1)
    }
}
```

****************************Claim capability****************************

The account A claims the capability published by account B.

```json
transaction {
    prepare(signer: AuthAccount) {
        let capability = signer.inbox.claim<&AuthAccount>("accountCapB", provider: 0x2)!
        let accountRef = capability.borrow()!
    }
}
```

## How is account linking useful?

Account linking is very useful in the case of onboarding non-existing Flow users to your Flow application. Because participating on Flow network requires having a Flow account, which a non-web3-native user will struggle to achieve a lot of applications create the account for the user.

> The problem with this approach becomes when a user wants to use that account with another application, or move the items stored in that account out of the application. With account linking it’s possible for the application to create the account instead of the user, but still allow for the user to later link that account to any other accounts he will own.

Reviewer 2: Want to avoid using terminology that implies a problem.

```suggestion
Naturally, users may expect to use their account with another application, or otherwise move assets stored in that account elsewhere. When an app initially leverages account linking it creates the account instead of the user and is where that user's specific state is stored. At a later point, users can take ownership of the app account providing they possess a full Flow account, typically by installing a wallet app.
```

> Account linking allows having multiple Flow accounts and abstracts them as a single account.

Reviewer 2: ```suggestion
Account linking enables users to possess multiple linked child accounts from different apps. Complexities associated with accessing those child accounts are eliminated by abstracting access to them through the user's parent account. Simply put, child accounts are accessed and treated as a seamless part of the parent account.
```

All assets in the app account can now jump the walled garden to play in the rest of the Flow ecosystem. No transactions are needed to move assets from the child account to the parent account, as the parent account already has full access to the assets in the child account.

![Multiple parent-child accounts on Flow](resources/multiple-accounts.png)

This shared control over the digital items in the in-app account enables users to establish real ownership of the items beyond the context of the app, where they can use their parent account to view inventory, take the items to other apps in the ecosystem, such as a marketplace or a game.

Most importantly, users are able to do this without the need to transfer the digital items between accounts, making it seamless to continue using the original app while also using other apps.

> ## Security Consideration with Account Linking

Reviewer: ```suggestion
## Security Considerations
```

> Account linking is a very powerful feature of Cadence and thus it must be treated with care. In this document, we’ve always talked about account linking between two accounts we own, even if the child account is managed by a third-party application, that we access through an authentication mechanism, it’s still treated as our account.

Reviewer: ```suggestion
Account linking is a _very_ powerful Cadence feature, and thus it must be treated with care. So far in this document, we’ve talked about account linking between two accounts we own, even if the child account is managed by a third-party application. But, we can't make the same trust assumptions about custodial accounts in the real world.
```

Creating an `AuthAccount` capability and publishing it to an account we don’t own means we are giving that account full access to our account. This should be seen as an anti-pattern.

<aside>

> ❗ Creating an `AuthAccount` capability and sharing it with another person account is in concept the same as giving that person your accounts private keys.

Reviewer: ```suggestion
❗ Creating an `AuthAccount` capability and sharing it with third-party account is in concept the same as giving that person your account's private keys.
```

</aside>

> Because pure account linking can be dangerous Flow introduces a Hybrid Custody contract that helps manage child accounts and access.

Reviewer: ```suggestion
Because pure account linking can be dangerous, Flow introduces the `HybridCustody` contract that helps custodial applications regulate access and parent account manage their many child accounts.
```

## Hybrid Custody and Account Linking

Apps need assurances that their own resources are safe from malicious actors, so giving out full access might not be the form they want. Using hybrid custody contracts, the app still maintains control of their managed accounts, but they can:

> 1. Share capabilities freely, with a few built-in controls over the types of capabilities that can be returned with some helper contracts (the `CapabilityFactory`, and `CapabilityFilter`)

Reviewer: ```suggestion
1. Share capabilities freely, with a few built-in controls over the types of capabilities that can be retrieved by parent accounts via helper contracts (the `CapabilityFactory`, and `CapabilityFilter`)
```

2. Share additional capabilities (public or private) with a parent account via a `CapabilityDelegator` resource

> Learn more about it in the Hybrid Custody documentation.

Reviewer 2: Link to the docs?

Author: Plan to make a final PR with all the links since not all the docs were there at the time of writing.
